Feature: flatFractions display flag

tpilewicz opened the pull request:
This addresses issue #626 

The flatFractions display flag is used to display fractions horizontally, it renders them as Latex: `\left. numerator \middle/ denominator \right.`.

Sometimes, the numerator or denominator contain other operations, and it's nice to display brackets around them to make it clear that for example in `(a + b)/4` the complete numerator is `a + b`. To identify these cases, I looked for a whitespace in the numerator and denominator. This solution might not be very elegant and although I've tested it with several situations, I might be forgetting some edge cases. Tell me what you think of it.

Also, my editor linted and removed trailing whitespaces in the modified files. I've put this in two separate commits, so that it is easy to revert if you'd rather not include this in the PR. I also advise you to read the PR commit by commit as this linting thing is hiding the main changes I've made.

christianp commented:
Thanks for this. I'm not keen on using the presence of spaces to decide whether to bracket - I'd rather keep a precise handle on when we _need_ brackets. I said in the issue that we could reuse the logic from `treeToJME`, but now I look at it, there's already a function `texifyOpArg` which adds brackets based on operator precedence, which we could use directly. So you'd return something like 

```
'\\left. ' + texifyOpArg(thing,texArgs,0) + ' \\middle/ ' + texifyOpArg(thing,texArgs,1) + ' \\right.'
```

I'm happy with removing the trailing whitespace - thanks for making it a separate commit!

christianp commented:
There are a few other places where fractions can be rendered - do a search/replace for `\frac` in jme-display.js. In particular, `texRationalNumber` needs to obey the new flag, and `diff` and `partialdiff` might as well do the same.

tpilewicz commented:
OK I'll add these changes in the coming days! Should be pretty easy if the function is already there :)

tpilewicz commented:
So I used `texifyOpArg` to apply brackets to numerator and denominator.

I added the flag in diff and partialdiff, always adding brackets to `texArgs[0]` in the case where `args[0]` is not of type 'name'. Otherwise for something like `diff(2x, y, 1)`, we would have `\left. \mathrm{d}2x \middle/ \mathrm{d}y \right.`, and `texifyOpArg` wouldn't change this.

I added the flag to `texRationalNumber`, but I'm not satisfied with result for mixedFractions that use a `mixedInteger`, it gives something like `1(\left. 1 \middle./ 2 \right.)` (for 1.5). Thought you might have ideas on how to prettify this? My personal opinion is that it will be hard to make flat fractions natural for the mixed fractions notation, I'd be in favor of not displaying flat fractions at all in that case.

christianp commented:
Thanks for this! I've merged it in, after making a little change: I think the best thing for the combination of mixedFractions and flatFractions is to put a space between the integer and fraction parts, rather than putting the fraction in brackets.

I've added your username to the CONTRIBUTORS.md file. Please put your real name in if you want.

tpilewicz commented:
Cool! Thanks for reviewing and helping on this :)
